### PR TITLE
sql: add disable_changefeed_replication session variable

### LINKED ISSUE #114071: sql: add session variable for CDC session-based filtering
Add a session variable (`disable_changefeed_replication`) that determines whether changes made during that session will persist through the changefeed. The default value will be `false`.

Jira issue: CRDB-33321

Epic CRDB-13169

### PR DESCRIPTION
**Pull request by andyyang890**
This patch adds a `disable_changefeed_replication` session variable
that can be used to disable changefeed replication for changes that
occur within a session. Right now, the session variable has no effect
but in later commits, it will be plumbed to the KV layer.

Fixes #114071

Release note: None

### CODE REVIEW

File: pkg/ccl/logictestccl/testdata/logic_test/changefeed
Comment (miretskiy) [inline pkg/ccl/logictestccl/testdata/logic_test/changefeed:77]: perhaps it might make sense to just invoke changefeed to make sure it doesn't crash when set to true?

**Review by miretskiy — APPROVED** [review]
:lgtm: from cdc side; but you should wait for real review from sql people.

File: pkg/ccl/logictestccl/testdata/logic_test/changefeed
Comment (andyyang890) [reply pkg/ccl/logictestccl/testdata/logic_test/changefeed:87]: Added!

**Review by yuzefovich — COMMENTED** [review]
> Reviewed 6 of 7 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/113992)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andyyang890, @mgartner, @miretskiy, and @rafiss)

___
*[`-- commits` line 7 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/113992#-Nil9H3m9WS9zL3SpEqM):*
How will it be propagated to KV layer? Will we be setting some new boolean field on the header of BatchRequest? Do CDC processors need to be aware of this? The reason I'm asking is to understand whether it's ok to keep it in `LocalOnlySessionData` or we need to include it into serialized `SessionData` that gets sent to remote nodes as part of processor protos.

**Review by andyyang890 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/113992)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @miraradeva, @miretskiy, and @rafiss)

___
*[`-- commits` line 7 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/113992#-Nil9H3m9WS9zL3SpEqM):*
My understanding is that when the session variable is set to true, the SQL layer will set a flag on any KV transactions that are created in that session, and then the KV layer will pass this flag on to the storage layer via the MVCCValueHeader, and then the rangefeeds will use that flag and not emit events for those value changes.

There are more details here if you're interested: https://docs.google.com/document/d/1bYCLrgOxb2yjTc5aWpZXl8dSB2w60F57GMUcehDRyzY/edit#heading=h.cbdzyq1u1gsg

> Do CDC processors need to be aware of this?

I think the answer is no because the CDC processors themselves won't be doing the filtering, rather the KV layer/rangefeeds won't even emit the changes to its subscribers.

@miretskiy, @miraradeva: would appreciate your input on whether this is right!

**Comment by miretskiy** [comment]
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/113992)_ status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @miraradeva, @miretskiy, and @rafiss)
> 
> _[`-- commits` line 7 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/113992#-Nil9H3m9WS9zL3SpEqM):_ My understanding is that when the session variable is set to true, the SQL layer will set a flag on any KV transactions that are created in that session, and then the KV layer will pass this flag on to the storage layer via the MVCCValueHeader, and then the rangefeeds will use that flag and not emit events for those value changes.
> 

Your understanding is correct.

> There are more details here if you're interested: https://docs.google.com/document/d/1bYCLrgOxb2yjTc5aWpZXl8dSB2w60F57GMUcehDRyzY/edit#heading=h.cbdzyq1u1gsg
> 
> > Do CDC processors need to be aware of this?
> 
> I think the answer is no because the CDC processors themselves won't be doing the filtering, rather the KV layer/rangefeeds won't even emit the changes to its subscribers.
> 
> @miretskiy, @miraradeva: would appreciate your input on whether this is right!

Yes, this sounds right -- cdc doesn't need anything (at this point) -- filtering happens on KV side (rangefeed processor)

**Review by miraradeva — COMMENTED** [review]
This all sounds right to me. On the KV side, there will be a boolean flag on the transaction proto, which will then be propagated on each write and persisted on the MVCCValueHeader (only when the flag is set). Here's the transaction proto change: https://github.com/cockroachdb/cockroach/pull/113522.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/113992#-:-NioIcGfDZOi2Luliosc:b-yw9vfm)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @miretskiy, and @rafiss)

**Review by yuzefovich — APPROVED** [review]
Thanks for the context. :lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/113992#-:-NioZJO0A-sUF5G6X_AF:bo9v8qd)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @miretskiy, and @rafiss)

**Comment by andyyang890** [comment]
TFTRs!

bors r=miretskiy,yuzefovich

**Comment by andyyang890** [comment]
blathers backport 23.2

**Pull request merged**
